rabit/src/allreduce_base.h

```diff
@@ -35,6 +37,22 @@ class Datatype {
 }
 namespace rabit {
 namespace engine {
+
+template <typename Fn, typename... Args>
+auto MakeTimeoutTask(Fn&& fn, int32_t timeout, int32_t rank) {
```

Comment:
The old time out implementation is incorrect.  Exception needs to  be thrown at main thread.